scripts/stac/imagery/generate_metadata.py

```diff
@@ -0,0 +1,181 @@
+from datetime import datetime
+from enum import Enum
+from typing import Optional
+
+
+class SubtypeParameterError(Exception):
+    def __init__(self, category: str) -> None:
+        self.message = f"Unrecognised/Unimplemented Subtype Parameter: {category}"
+
+
+class ImagerySubtypes(str, Enum):
+    SATELLITE = "Satellite Imagery"
+    URBAN = "Urban Aerial Photos"
+    RURAL = "Rural Aerial Photos"
+    AERIAL = "Aerial Photos"
+    HISTORICAL = "Scanned Aerial Photos"
+
+
+class ElevationSubtypes(str, Enum):
+    DEM = "DEM"
+    DSM = "DSM"
+
+
+HUMAN_READABLE_REGIONS = {
+    "antarctica": "Antarctica",
+    "auckland": "Auckland",
+    "bay-of-plenty": "Bay of Plenty",
+    "canterbury": "Canterbury",
+    "gisborne": "Gisborne",
+    "global": "Global",
+    "hawkes-bay": "Hawke's Bay",
+    "manawatu-whanganui": "Manawatū-Whanganui",
+    "marlborough": "Marlborough",
+    "nelson": "Nelson",
+    "new-zealand": "New Zealand",
+    "northland": "Northland",
+    "otago": "Otago",
+    "pacific-islands": "Pacific Islands",
+    "southland": "Southland",
+    "taranaki": "Taranaki",
+    "tasman": "Tasman",
+    "waikato": "Waikato",
+    "wellington": "Wellington",
+    "west-coast": "West Coast",
+}
+
+
+def generate_title(
+    category: str,
+    region: str,
+    gsd: str,
+    start_datetime: datetime,
+    end_datetime: datetime,
+    lifecycle: str,
+    location: Optional[str] = None,
+    event: Optional[str] = None,
+    historic_survey_number: Optional[str] = None,
+) -> str:
+    """Generates the title for imagery and elevation datasets.
+    Satellite Imagery / Urban Aerial Photos / Rural Aerial Photos:
+    [Location Name / Region if no Location Name specified] [GSD] [?Event Name] [Data Sub-Type] ([Year(s)]) [?- Preview]
+    DEM / DSM:
+    [Location Name / Region if no Location Name specified] [?- Event Name] LiDAR [GSD] [Data Sub-Type] ([Year(s)]) [?- Preview]
+    If Historic Survey Number:
+    [Location Name / Region if no Location Name specified] [GSD] [Survey Number] ([Year(s)]) [?- Preview]
+
+    Args:
+        category: Dataset category description - used to determine if the dataset is imagery, elevation, ect.
+        region: Region of Dataset
+        gsd: Dataset Ground Sample Distance
+        start_date: Dataset capture start date
+        end_date: Dataset capture end date
+        lifecycle: Dataset status
+        location: Optional location of dataset, e.g.- Hutt City
+        event: Optional details of capture event, e.g. - Cyclone Gabreille
+        historic_survey_number: Optional historic imagery survey number, e.g.- SNC88445
+    Returns:
+        Dataset Title
+    """
+    # pylint: disable-msg=too-many-arguments
```

Comment:
What is the pythonic way of handling too many arguments especially with similar "str" arguments it can be very easy to put them into the wrong order eg swap region for gsd?

Reply:
Adding `*` as a pseudo-parameter before all the others force users to pass arguments by name, making it much less likely to accidentally swap them. This is already considered best practice by some projects, since it eliminates a whole class of problems with very little overhead. From the [docs](https://docs.python.org/3.5/reference/compound_stmts.html#function-definitions):

> Parameters after “`*`” or “*`identifier`” are keyword-only parameters and may only be passed used keyword arguments.

Reply:
I have turned the optional args into **kwargs, this was enough to remove the pylint warning.
I left the required args as is because we want the function to fail if they are not provided?

Reply:
I'd rather revert and disable too-many-arguments globally than use this pattern - this is hiding information inside `kwargs`.

Reply:
That said, doesn't the `generate_title` and `generate_description` logic belong inside `ImageryCollection`?

Reply:
Sorry Victor somehow I missed your first comment I have added the pseudo * parameter.

I've left the kwargs but can change it back, seems like we have to pick and I'm not sure what the best/right option is..

I split this code into its own file as the collection file was already quite long but I can merge them together if that is preferred?

Reply:
`**kwargs` is appropriate if you don't know which arguments the function is going to be called with - you can't know what the parameter names should be. But this is not the case here - you know which keyword arguments are possible, and only ever handle those specifically. Which means the function signature is less strict than it could be.

Reply:
As discussed, *if* `generate_title` and `generate_description` are always run in the context of trying to create an `ImageryCollection`, it probably makes sense to instead pass their parameters to the `ImageryCollection` constructor, then *either* have the constructor set `self.title` and `self.description` itself, or have `@property`-annotated methods for both of them. Creating the title and description directly in the constructor (possibly via private methods) makes sense if the constructor receives all the information necessary to create them, that is, there's nothing which should ever modify the title or description outside of the constructor. Using `@property` methods make more sense if anything outside the constructor could cause the title or description to change.

Reply:
So I've:
- put all the user inputs into a typed dict which gets rid of the pylint warning and removes the kwargs
- Moved title and description into the ImageryCollection class as private methods, at this stage I don't expect these to change after being created so opted for that over `@property`